test.py

```diff
@@ -517,53 +517,78 @@ def __init__(self, test_no, shortname, suite):
             "--input={}".format(self.cql),
             "--output={}".format(self.tmpfile),
         ]
+        self.is_pre_check_ok = False
         self.is_executed_ok = False
         self.is_new = False
+        self.is_post_check_ok = False
         self.is_equal_result = None
         self.summary = "not run"
         self.unidiff = None
+        self.server_log = None
         self.env = dict()
 
     async def run(self, options):
-        async with self.suite.clusters.instance() as cluster:
-            self.args.insert(1, "--host={}".format(cluster[0].host))
-            cluster[0].take_log_savepoint()
-            self.is_executed_ok = await run_test(self, options, env=self.env)
-            if self.is_executed_ok is False:
-                self.server_log = cluster[0].read_log()
-
         self.success = False
         self.summary = "failed"
 
         def set_summary(summary):
             self.summary = summary
             logging.info("Test %d %s", self.id, summary)
+            if self.server_log:
+                logging.info("Server log:\n%s", self.server_log)
 
-        if not os.path.isfile(self.tmpfile):
-            set_summary("failed: no output file")
-        elif not os.path.isfile(self.result):
-            set_summary("failed: no result file")
-            self.is_new = True
-        else:
-            self.is_equal_result = filecmp.cmp(self.result, self.tmpfile)
-            if self.is_equal_result is False:
-                self.unidiff = format_unidiff(self.result, self.reject)
-                set_summary("failed: test output does not match expected result")
-                logging.info("\n{}".format(palette.nocolor(self.unidiff)))
-            elif self.is_executed_ok:
-                self.success = True
-                set_summary("succeeded")
-            else:
-                set_summary("failed: correct output but non-zero return status.\nCheck test log.")
-
-        if self.is_new or self.is_equal_result is False:
-            # Put a copy of the .reject file close to the .result file
-            # so that it's easy to analyze the diff or overwrite .result
-            # with .reject. Preserve the original .reject file: in
-            # multiple modes the copy .reject file may be overwritten.
-            shutil.copyfile(self.tmpfile, self.reject)
-        elif os.path.exists(self.tmpfile):
-            pathlib.Path(self.tmpfile).unlink()
+        async with self.suite.clusters.instance() as cluster:
+            self.args.insert(1, "--host={}".format(cluster[0].host))
+            # If pre-check fails, e.g. because Scylla failed to start
+            # or crashed between two tests, fail entire test.py
```

Comment:
If Scylla fails between two tests, it's enough to do a "post check" (right after a test finished, were you can also report which test is to blame). You don't really need to do a "pre-check" yet again a millisecond later when starting the next test.

You only need to check that Scylla booted correctly when starting Scylla, not before each test.

Unless I'm misunderstanding what you're trying to do here.

Reply:
I agree we could optimize out getting the number of keyspaces - we don't need to get it each time in the beginning of the test. pre_check(), however, also prints a line into the log file with the test name. I think I will rename pre_check to before_test and post_check to after_test(), these names would better reflect the responsibility of these two calls.

Finally, you maybe are raising in this comment the issue of when to throw if the server fails to start. I thought that it is perhaps better to throw from pre_check (before_test()) since then the test can consider what to do with it. If building a new server throws, i.e.. when the server is being added to the pool, I'll have to handle this twice - first when the server fails to start, and then when it crashes after the test.

To sum up:
- I will rename the two functions
- I will optimize querying the canonical keyspace count so that it's done once per server, not once for each test.
- I will keep the before- function 
- I will keep suppressing the start exception and will raise it in the before- function.

Please let me know what you think